packages/middleware-bucket-endpoint/src/bucketEndpointMiddleware.ts

```diff
@@ -0,0 +1,76 @@
+import { bucketHostname } from "./bucketHostname";
+import { BucketEndpointResolvedConfig } from "./configurations";
+import {
+  BuildHandler,
+  BuildHandlerArguments,
+  BuildHandlerOptions,
+  BuildHandlerOutput,
+  BuildMiddleware,
+  MetadataBearer,
+  Pluggable
+} from "@aws-sdk/types";
+import { HttpRequest } from "@aws-sdk/protocol-http";
+
+export function bucketEndpointMiddleware(
+  options: BucketEndpointResolvedConfig
+): BuildMiddleware<any, any> {
+  return <Output extends MetadataBearer>(
+    next: BuildHandler<any, Output>
+  ): BuildHandler<any, Output> => async (
+    args: BuildHandlerArguments<any>
+  ): Promise<BuildHandlerOutput<Output>> => {
+    const {
+      Bucket: bucketName,
+      $bucketEndpoint,
+      $forcePathStyle = options.forcePathStyle,
+      $useAccelerateEndpoint = options.useAccelerateEndpoint,
+      $useDualstackEndpoint = options.useDualstackEndpoint
```

Comment:
Maybe I missed it somewhere, where are these `$` keys added to generated input types(shapes)?

Reply:
Regarding `$useDualstackEndpoint` and `$useAccelerateEndpoint`, if s3 decide to add them as regions to endpoints.json, we will in a bad place.

Reply:
I think we should remove these $ input options.  The modular command config provides enough flexibility.

Reply:
removed in e1426ddc93cb3af5d2c5dc87d5bfd5993cb62f53